File: backend/app/routes/categories.py
from flask import Blueprint, jsonify, request
from app import db
from app.models import Category

categories_bp = Blueprint('categories', __name__, url_prefix='/api/categories')

@categories_bp.route('/', methods=['GET'])
def get_categories():
    """Récupérer toutes les catégories triées par ordre d'affichage"""
    categories = Category.query.order_by(Category.display_order).all()
    return jsonify([c.to_dict() for c in categories]), 200

@categories_bp.route('/', methods=['POST'])
def create_or_update_categories():
    """Créer une nouvelle catégorie ou réinitialiser les catégories"""
    data = request.get_json()
    
    try:
        # Si c'est une liste, reorder tous (bulk update)
        if isinstance(data, list):
            for item in data:
                category = Category.query.filter_by(name=item['name']).first()
                if category:
                    category.display_order = item.get('display_order', 999)
                    category.is_visible = item.get('is_visible', True)
                else:
                    category = Category(
                        name=item['name'],
                        display_order=item.get('display_order', 999),
                        is_visible=item.get('is_visible', True)
                    )
                    db.session.add(category)
            db.session.commit()
            return jsonify({'message': 'Catégories mises à jour'}), 200
        # Si c'est un objet, créer une nouvelle catégorie
        elif isinstance(data, dict):
            # Vérifier si la catégorie existe déjà
            if 'name' not in data:
                return jsonify({'error': 'Le nom est requis'}), 400
            
            existing = Category.query.filter_by(name=data['name'].lower()).first()
            if existing:
                return jsonify({'error': f'Catégorie "{data["name"]}" existe déjà'}), 400
            
            category = Category(
                name=data['name'].lower(),
                display_order=data.get('display_order', 999),
                is_visible=data.get('is_visible', True)
            )
            db.session.add(category)
            db.session.commit()
            return jsonify(category.to_dict()), 201
        else:
            return jsonify({'error': 'Format invalide'}), 400
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@categories_bp.route('/<int:category_id>', methods=['PUT'])
def update_category(category_id):
    """Mettre à jour une catégorie"""
    category = Category.query.get(category_id)
    if not category:
        return jsonify({'error': 'Catégorie non trouvée'}), 404
    
    data = request.get_json()
    if 'display_order' in data:
        category.display_order = data['display_order']
    if 'is_visible' in data:
        category.is_visible = data['is_visible']
    
    db.session.commit()
    return jsonify(category.to_dict()), 200

@categories_bp.route('/init', methods=['POST'])
def init_categories():
    """Initialiser les 4 catégories par défaut"""
    try:
        defaults = [
            {'name': 'classique', 'display_order': 1},
            {'name': 'gourmand', 'display_order': 2},
            {'name': 'sain', 'display_order': 3},
            {'name': 'special', 'display_order': 4}
        ]
        
        for cat in defaults:
            existing = Category.query.filter_by(name=cat['name']).first()
            if not existing:
                category = Category(name=cat['name'], display_order=cat['display_order'], is_visible=True)
                db.session.add(category)
        
        db.session.commit()
        return jsonify({'message': 'Catégories initialisées'}), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

File: backend/routes/categories.js
const router = require('express').Router();
const db = require('../db');

router.get('/', async (req, res) => {
  try {
    const { rows } = await db.query('SELECT * FROM categories ORDER BY display_order');
    res.json(rows);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/', async (req, res) => {
  const data = req.body;
  try {
    if (Array.isArray(data)) {
      for (const item of data) {
        const { rows } = await db.query('SELECT id FROM categories WHERE name=$1', [item.name]);
        if (rows[0]) {
          await db.query('UPDATE categories SET display_order=$1, is_visible=$2 WHERE name=$3',
            [item.display_order ?? 999, item.is_visible ?? true, item.name]);
        } else {
          await db.query('INSERT INTO categories (name, display_order, is_visible) VALUES ($1,$2,$3)',
            [item.name, item.display_order ?? 999, item.is_visible ?? true]);
        }
      }
      return res.json({ message: 'Catégories mises à jour' });
    }
    if (!data.name) return res.status(400).json({ error: 'Le nom est requis' });
    const name = data.name.toLowerCase();
    const { rows } = await db.query('SELECT id FROM categories WHERE name=$1', [name]);
    if (rows[0]) return res.status(400).json({ error: `Catégorie "${name}" existe déjà` });
    const { rows: [cat] } = await db.query(
      'INSERT INTO categories (name, display_order, is_visible) VALUES ($1,$2,$3) RETURNING *',
      [name, data.display_order ?? 999, data.is_visible ?? true]
    );
    res.status(201).json(cat);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.put('/:id', async (req, res) => {
  try {
    const { rows: existing } = await db.query('SELECT * FROM categories WHERE id=$1', [req.params.id]);
    if (!existing[0]) return res.status(404).json({ error: 'Catégorie non trouvée' });
    const c = existing[0];
    const { display_order=c.display_order, is_visible=c.is_visible } = req.body;
    const { rows } = await db.query(
      'UPDATE categories SET display_order=$1, is_visible=$2 WHERE id=$3 RETURNING *',
      [display_order, is_visible, req.params.id]
    );
    res.json(rows[0]);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/init', async (req, res) => {
  const defaults = [
    { name: 'classique', display_order: 1 },
    { name: 'gourmand',  display_order: 2 },
    { name: 'sain',      display_order: 3 },
    { name: 'special',   display_order: 4 }
  ];
  try {
    for (const cat of defaults) {
      const { rows } = await db.query('SELECT id FROM categories WHERE name=$1', [cat.name]);
      if (!rows[0]) {
        await db.query('INSERT INTO categories (name, display_order, is_visible) VALUES ($1,$2,true)',
          [cat.name, cat.display_order]);
      }
    }
    res.json({ message: 'Catégories initialisées' });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

module.exports = router;
